zkvm/src/predicate.rs

```diff
@@ -88,46 +95,31 @@ impl Predicate {
         }
     }
 
-    /// Creates a predicate with witness being an opaque branch of the tree.
-    pub fn opaque_branch(point: CompressedRistretto) -> Result<Self, VMError> {
-        Ok(Predicate::Witness(PredicateWitness::OpaqueBranch(
-            point.decompress().ok_or(VMError::FormatError)?,
-        )))
-    }
-
-    /// Creates a predicate with a signing key witness.
-    pub fn from_signing_key(secret_key: Scalar) -> Self {
-        Predicate::Witness(PredicateWitness::Key(secret_key).into())
-    }
-
     /// Downcasts the predicate to a signing key
-    pub fn to_signing_key(self) -> Result<Scalar, VMError> {
-        match self.to_witness()? {
-            PredicateWitness::Key(s) => Ok(s),
+    pub fn to_key(self) -> Result<VerificationKey, VMError> {
+        match self {
+            Predicate::Key(k) => Ok(k),
             _ => Err(VMError::TypeNotKey),
         }
     }
 
     /// Creates a disjunction of two predicates.
     pub fn or(self, right: Predicate) -> Result<Self, VMError> {
-        let l = self.to_witness()?;
-        let r = right.to_witness()?;
-        Ok(Predicate::Witness(
-            PredicateWitness::Or(Box::new(l), Box::new(r)).into(),
-        ))
+        let point = {
+            let l = self.to_point();
+            let f = Predicate::commit_or(l, right.to_point());
+            l.decompress().ok_or(VMError::InvalidPoint)? + f * PedersenGens::default().B
```

Comment:
maybe make this line a bit clearer?

```suggestion
            let l = l.decompress().ok_or(VMError::InvalidPoint)?;
            l + f * PedersenGens::default().B
```